### PR TITLE
feat:remove create-key circular dependency

### PR DESCRIPTION
**Pull request by aroralanuk**
### Description

Diagnosis: circular dependency created by igp requiring multisig size and threshold for building coreConfig which is needed while creating new keys. This arises in the scenario where you add the chain to chainMetadata and not to defaultMultisigConfigs (say newChainConfigs)

- separated agents and helloworld configs from core and segregated scripts for `core-utils.ts` and `agent-utils.ts` for avoiding loading of envConfig.coreConfig
- add argv commands to prompt for new chains or updated chains and validator/threshold combo:
`yarn ts-node scripts/create-keys -e testnet4 --context hyperlane --new-chains basesepolia=2/3` (here 2 is desired threshold and 3 is desired multisig set size) 
  - if you provide a newChainConfigs chain and don't supply an argv, the create-keys script will revert
- persist addresses in gcp (already present) and under sdk (this PR) with types preserved

### Drive-by changes

- imports
- might break other scripts like delete-keys, etc in infra (not tested)

### Related issues

- fixes https://github.com/hyperlane-xyz/issues/issues/887

### Backward compatibility

Yes

### Testing

Manual

### CODE REVIEW

File: typescript/infra/scripts/agent-utils.ts [hunk 1]
```diff
@@ -113,6 +114,14 @@ export function withKeyRoleAndChain<T>(args: yargs.Argv<T>) {
     .alias('i', 'index');
 }
 
+// missing chains are chains needed which are not as part of defaultMultisigConfigs in sdk/src/consts/ but are in chainMetadata
+export function withMissingChains<T>(args: yargs.Argv<T>) {
+  return args
+    .describe('new-chains', 'new chains to add')
+    .string('new-chains')
+    .alias('n', 'new-chains');
+}
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:123]: this is oddly specific
doesnt chainMetadata also require SDK changes?

File: typescript/infra/scripts/agent-utils.ts [hunk 2]
```diff
@@ -122,39 +131,100 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
-export function getEnvironmentConfig(environment: DeployEnvironment) {
-  return environments[environment];
-}
-
-export async function getConfigsBasedOnArgs(argv?: {
+// not requiring to build coreConfig to get agentConfig
+export async function getAgentConfigsBasedOnArgs(argv?: {
   environment: DeployEnvironment;
   context: Contexts;
+  'new-chains': string;
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:138]: I dont understand this syntax

File: typescript/infra/scripts/agent-utils.ts [hunk 3]
```diff
@@ -122,39 +131,100 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+
+  const newValidatorCounts: ChainMap<number> = {};
+  const newThresholds: ChainMap<number> = {};
+  if (newChains) {
+    const chains = newChains.split(',');
+    for (const chain of chains) {
+      const [chainName, threshold] = chain.split('=');
+      const [newThreshold, newValidatorCount] = threshold.split('/');
+      newThresholds[chainName] = Number(newThreshold);
+      newValidatorCounts[chainName] = Number(newValidatorCount);
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:154]: maybe parseInt?

Comment (yorhodes) [reply typescript/infra/scripts/agent-utils.ts:154]: ngl this feels pretty fragile
idk if a regex is better?

File: typescript/infra/scripts/agent-utils.ts [hunk 4]
```diff
@@ -122,39 +131,100 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+  }
+
+  const agentConfig = getAgentConfig(context, environment);
+  // check if new chains are needed
+  const missingChains = checkIfValidatorsArePresisted(agentConfig);
+
+  for (const chain of missingChains) {
+    if (!Object.keys(newValidatorCounts).includes(chain)) {
+      throw new Error(`Missing chain ${chain} not specified in new-chains`);
+    }
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:165]: I dont understand this failure mode

File: typescript/infra/scripts/agent-utils.ts [hunk 5]
```diff
@@ -122,39 +131,100 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+      validators,
+    };
+  }
+
+  return {
+    agentConfig,
+    context,
+    environment,
+    newThresholds,
+  };
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:198]: why do we need to return all of this?

File: typescript/infra/scripts/agent-utils.ts [hunk 6]
```diff
@@ -122,39 +131,100 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+    throw new Error(
+      `Context ${context} does not exist in agents for environment ${environment}`,
     );
-  return agentConfig;
+  }
+  return agentsForEnvironment[context];
+}
+
+// check if validators are persisted in agentConfig
+export function checkIfValidatorsArePresisted(
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:220]: nit: persisted

File: typescript/infra/config/environments/agents.ts
Comment (yorhodes) [inline typescript/infra/config/environments/agents.ts:1]: seems like a step in the wrong direction from deprecating infra?

File: typescript/infra/scripts/core-utils.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+import { getAgentConfig, getArgs, withContext } from './agent-utils';
+
+// utils which use both environment configs
+
+export function getEnvironmentConfig(environment: DeployEnvironment) {
+  return environments[environment];
+}
+
+export async function getConfigsBasedOnArgs(argv?: {
```
Comment (yorhodes) [inline typescript/infra/scripts/core-utils.ts:13]: why move this and have the diff be so big?

File: typescript/infra/scripts/create-keys.ts [hunk 8]
```diff
@@ -1,10 +1,11 @@
...
 
-import { getConfigsBasedOnArgs } from './utils';
+import { getAgentConfigsBasedOnArgs } from './agent-utils';
 
 async function main() {
-  const { agentConfig } = await getConfigsBasedOnArgs();
-  return createAgentKeysIfNotExists(agentConfig);
+  const { agentConfig, newThresholds } = await getAgentConfigsBasedOnArgs();
+  await createAgentKeysIfNotExists(agentConfig, newThresholds);
+  return 'Keys created successfully!';
```
Comment (yorhodes) [inline typescript/infra/scripts/create-keys.ts:8]: nice

**Review by yorhodes — COMMENTED** [review]
it seems feasible to not have such a large diff simply by avoiding moving stuff around
that makes your reviewer's life easier
https://mtlynch.io/code-review-love/

**Comment by aroralanuk** [comment]
> * separated agents and helloworld configs from core and segregated scripts for `core-utils.ts` and `agent-utils.ts` for avoiding loading of envConfig.coreConfig

from the PR description: "separated agents and helloworld configs from core and segregated scripts for core-utils.ts and agent-utils.ts for avoiding loading of envConfig.coreConfig" other the create-key doesn't work.

envConfig gets loaded -> builds the core config -> builds igp config -> calls multisigIsmVerificationCost -> relies on the validator being present in the multisig list

File: typescript/infra/config/environments/agents.ts
Comment (aroralanuk) [reply typescript/infra/config/environments/agents.ts:1]: the goal is to deduplicate between the CLI and the infra but maintain agent specific code in the infra.

File: typescript/infra/scripts/agent-utils.ts (hunk 1 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:123]: oddly specific how, you want to create keys for new chains which I don't have in the SDK so I'll ask for the threshold and total # of keys.

File: typescript/infra/scripts/agent-utils.ts (hunk 2 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:138]: see above

File: typescript/infra/scripts/agent-utils.ts (hunk 3 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:154]: This is similar to the parseContextAndRoles for the fund-keys-from-deployer script

File: typescript/infra/scripts/agent-utils.ts (hunk 4 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:165]: "if you include a chain in chainMetadata but not in the multisig.json, you need to specify the new chain in new-chains" added this as a comment

File: typescript/infra/scripts/agent-utils.ts (hunk 5 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:198]: We were already returning most of it except newThresholds which is needed for createAgentKeysIfNotExists

File: typescript/infra/scripts/core-utils.ts (hunk 7 above)
Comment (aroralanuk) [reply typescript/infra/scripts/core-utils.ts:13]: see above

File: typescript/infra/scripts/agent-utils.ts (hunk 6 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:220]: done

File: typescript/infra/scripts/agent-utils.ts (hunk 1 above)
Comment (aroralanuk) [reply typescript/infra/scripts/agent-utils.ts:123]: updated to the total number of keys

File: typescript/infra/config/kathy.json
Comment (yorhodes) [inline typescript/infra/config/kathy.json:1]: just curious, do we need entries for every environment and context even when non exist?

File: typescript/infra/scripts/agent-utils.ts [hunk 9]
```diff
@@ -132,45 +141,97 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+          () => '0x0',
+        ),
+      },
+      chain as Chains,
+    );
+    // the hardcoded fields are not strictly necessary to be accurate for create-keys.ts
+    // ideally would still get them from the chainMetadata
+    agentConfig.validators!.chains[chain] = {
+      interval: 1, // chainMetadata.estimateBlockNumber is optional
+      reorgPeriod: 0, // chainMetadata.reorgPeriod is optional
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:196]:
```suggestion
      interval: chainMetadata.estimateBlockNumber ?? 1,
      reorgPeriod: chainMetadata.reorgPeriod ?? 0,
```

File: typescript/infra/scripts/agent-utils.ts [hunk 10]
```diff
@@ -132,45 +141,97 @@ export function assertEnvironment(env: string): DeployEnvironment {
...
+        [context]: Array.from(
+          { length: newValidatorCounts[chain] },
+          () => '0x0',
+        ),
+      },
+      chain as Chains,
+    );
+    // the hardcoded fields are not strictly necessary to be accurate for create-keys.ts
+    // ideally would still get them from the chainMetadata
+    agentConfig.validators!.chains[chain] = {
```
Comment (yorhodes) [inline typescript/infra/scripts/agent-utils.ts:194]: why do you need this `!`?

File: typescript/infra/src/agents/key-utils.ts [hunk 11]
```diff
@@ -352,52 +364,101 @@ export async function rotateKey(
...
+  fs.writeFileSync(filePath, JSON.stringify(addresses, null, 2));
 }
 
-async function fetchGCPKeyAddresses(
+// maintaining the multisigIsm schema sans threshold
+export async function persistValidatorAddressesToLocalArtifacts(
+  fetchedValidatorAddresses: ChainMap<{ validators: Address[] }>,
+) {
+  for (const chain of Object.keys(fetchedValidatorAddresses)) {
+    awMultsigAddresses[chain] = {
```
Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:438]: nit: awMultisigAddresses

File: typescript/infra/src/agents/key-utils.ts [hunk 12]
```diff
@@ -352,52 +364,101 @@ export async function rotateKey(
...
+      validators: fetchedValidatorAddresses[chain].validators, // fresh from aws
+    };
+  }
+  // Resolve the relative path
+  const filePath = path.resolve(__dirname, '../../config/aw-multisig.json');
+  // Write the updated object back to the file
+  fs.writeFileSync(filePath, JSON.stringify(awMultsigAddresses, null, 2));
```
Comment (yorhodes) [inline typescript/infra/src/agents/key-utils.ts:445]: nit: we have read and write json helpers

**Review by yorhodes — APPROVED** [review]

File: typescript/infra/config/kathy.json
Comment (aroralanuk) [reply typescript/infra/config/kathy.json:1]: yes, it's type checked by role, context and env.

**Pull request merged**